source/loading.c

```diff
@@ -570,11 +500,57 @@ bool load_preview(Entry_List_s list, C2D_Image * preview_image, int * preview_of
     char *preview_buffer = NULL;
     u64 size = load_data("/preview.png", entry, &preview_buffer);
 
-    if(!size)
+    if(size)
+    {
+        if (!(size = png_to_abgr(&preview_buffer, size)))
+        {
+            return false;
+        }
+    }
+    else
     {
         free(preview_buffer);
-        throw_error("No preview found.", ERROR_LEVEL_WARNING);
-        return false;
+
+        const int top_size =  TOP_SCREEN_WIDTH * SCREEN_HEIGHT * 4;
+        const int out_size = top_size * 2;
+
+        char * rgba_buffer = malloc(out_size);
+        memset(rgba_buffer, 0, out_size);
+        bool found_splash = false;
+
+        // try to assembly a preview from the splash screens
+        size = load_data("/splash.bin", entry, &preview_buffer);
+        if (size)
+        {
+            found_splash = true;
+            bin_to_abgr(&preview_buffer, size);
+
+            memcpy(rgba_buffer, preview_buffer, top_size);
+            free(preview_buffer);
+        }
+
+        size = load_data("/splashbottom.bin", entry, &preview_buffer);
+        if (size)
+        {
+            found_splash = true;
+            bin_to_abgr(&preview_buffer, size);
+
+            // Store the bottom splash screen under the top splash and centered
+            for (int i = 0; i < SCREEN_HEIGHT; ++i)
+                memcpy(rgba_buffer + top_size + ( TOP_SCREEN_WIDTH * SCREEN_COLOR_DEPTH * i) + ((TOP_SCREEN_WIDTH - BOTTOM_SCREEN_WIDTH) / 2 * SCREEN_COLOR_DEPTH), preview_buffer + (BOTTOM_SCREEN_WIDTH * SCREEN_COLOR_DEPTH * i), BOTTOM_SCREEN_WIDTH * SCREEN_COLOR_DEPTH);
```

Comment:
Can we break the calculation of each parameter for memcpy into easier to read steps rather than doing it all inline?

Reply:
a0eebc745905ac5f61d1bec0e23cac79e5a3438f puts the memcpy over multiple lines to make it easier to read
7a705f80b9b9c05a7773fed460d141df03fdb648 splits the logic into multiple constants